hoomd/md/EvaluatorPairTable.h

```diff
@@ -0,0 +1,236 @@
+                                                   free_F);
+                v["width"] = width;
+                v["r_min"] = rmin;
+                return v;
+                }
+            #endif
+            }
+            #ifdef SINGLE_PRECISION
+            __attribute__((aligned(8)));
+            #else
+            __attribute__((aligned(16)));
+            #endif
+
+
+        //! Constructs the pair potential evaluator
+        /*! \param _rsq Squared distance between the particles
+            \param _rcutsq Squared distance at which the potential goes to 0
+            \param _params Per type pair parameters of this potential
+        */
+        DEVICE EvaluatorPairTable(Scalar _rsq, Scalar _rcutsq, const param_type& _params)
+            : rsq(_rsq), rcutsq(_rcutsq), width(_params.width), rmin(_params.rmin), V_table(_params.V_table), F_table(_params.F_table)
+            {
+            }
+
+        //! Table doesn't use diameter
+        DEVICE static bool needsDiameter() { return false; }
+        //! Accept the optional diameter values
+        /*! \param di Diameter of particle i
+            \param dj Diameter of particle j
+        */
+        DEVICE void setDiameter(Scalar di, Scalar dj) { }
+
+        //! Table doesn't use charge
+        DEVICE static bool needsCharge() { return false; }
+        //! Accept the optional diameter values
+        /*! \param qi Charge of particle i
+            \param qj Charge of particle j
+        */
+        DEVICE void setCharge(Scalar qi, Scalar qj) { }
+
+        //! Evaluate the force and energy
+        /*! \param force_divr Output parameter to write the computed force divided by r.
+            \param pair_eng Output parameter to write the computed pair energy
+            \param energy_shift If true, the potential must be shifted so that
+            V(r) is continuous at the cutoff
+            \note There is no need to check if rsq < rcutsq in this method.
+            Cutoff tests are performed in PotentialPair.
+
+            \return True if they are evaluated or false if they are not because
+            we are beyond the cutoff
+        */
+        DEVICE bool evalForceAndEnergy(Scalar& force_divr, Scalar& pair_eng, bool energy_shift)
+            {
+            // compute the force divided by r in force_divr
+            if (rsq < rcutsq && std::sqrt(rsq) >= rmin)
+                {
+                Scalar delta_r = (std::sqrt(rcutsq) - rmin) / width;
+                // precomputed term
+                Scalar value_f = (std::sqrt(rcutsq) - std::sqrt(rsq)) / delta_r;
+
+                // compute index into the table and read in values
+                unsigned int value_i = (unsigned int)floor(value_f);
+                // unpack the data
+                Scalar V0 = V_table[value_i];
+                Scalar V1 = V_table[value_i + 1];
+                Scalar F0 = F_table[value_i];
+                Scalar F1 = F_table[value_i + 1];
```

Comment:
Handle out of bounds memory accesses:
```suggestion
                Scalar V0 = V_table[value_i];
                Scalar V1 = 0;
                if (value_i + 1 < width)
                    V1 = V_table[value_i + 1];
```

We might also be concerned about underflow or overflow of value_i for some unexpected floating point rounding conditions. We could write this more defensively by always computing `value_f` and `value_i` (as an `int`), then change the `if (rsq ...)` condition to `if (value_i >= 0 && value_i < width)`